File: Java/src/_0000_study/_leetcode/_lcp/_LCP_04_Broken_Board_Dominoes/Solution.java
package _0000_study._leetcode._lcp._LCP_04_Broken_Board_Dominoes;

import java.util.Arrays;

/**
 * Date: 2021/6/7 15:25
 */
public class Solution {

    int n, m;
    int[][] g;
    int[] match;
    boolean[] st;

    static int[] dx = {-1, 0, 1, 0}, dy = {0, 1, 0, -1};

    public int domino(int _n, int _m, int[][] broken) {
        n = _n; m = _m;
        g = new int[n][m];  // 1表示坏掉的格子
        match = new int[n * m];
        st = new boolean[n * m];
        for (int i = 0; i < broken.length; i++) {
            int a = broken[i][0], b = broken[i][1];
            g[a][b] = 1;
        }

        Arrays.fill(match, -1);
        int res = 0;
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                if ((i + j) % 2 != 0 && g[i][j] != 1) {
                    Arrays.fill(st, false);
                    if (find(i * m + j)) res++;
                }
        return res;
    }

    private boolean find(int x) {
        for (int i = 0; i < 4; i++) {
            int a = x / m + dx[i], b = x % m + dy[i];
            if (a < 0 || a >= n || b < 0 || b >= m || g[a][b] == 1) continue;

            int j = a * m + b;
            if (!st[j]) {
                st[j] = true;
                // 如果女孩j没有男朋友，或者她原来的男朋友能够预定其它喜欢的女孩。配对成功
                if (match[j] == -1 || find(match[j])) {
                    match[j] = x;
                    return true;
                }
            }
        }
        return false;
    }

    public static void main(String[] args) {

        int[][] broken = {{1, 0}, {1, 1}};
        System.out.println((new Solution()).domino(2, 3, broken));  // 2
    }
}

File: Python/_0000_study/_leetcode/_lcp/_LCP_04_Broken_Board_Dominoes/Solution.py
# coding=utf-8
# Date: 2021/6/7 15:44
from typing import List


# 执行用时：44 ms, 在所有 Python3 提交中击败了75.00%的用户
# 内存消耗：15.1 MB, 在所有 Python3 提交中击败了56.25%的用户
class Solution:
    def __init__(self):
        self.m = 0
        self.n = 0
        self.g = []  # 坏的为True
        self.match = []
        self.st = []
        self.dx = [-1, 0, 1, 0]
        self.dy = [0, 1, 0, -1]

    def domino(self, _n: int, _m: int, broken: List[List[int]]) -> int:
        self.m = _m
        self.n = _n
        self.g = [[False for _ in range(self.m)] for _ in range(self.n)]
        self.match = [-1 for _ in range(self.n * self.m)]
        self.st = [False for _ in range(self.n * self.m)]
        for i in range(len(broken)):
            a = broken[i][0]; b = broken[i][1]
            self.g[a][b] = True

        res = 0
        for i in range(self.n):
            for j in range(self.m):
                if (i + j) % 2 != 0 and not self.g[i][j]:
                    self.st = [False for _ in range(self.n * self.m)]
                    if self.find(i * self.m + j):
                        res += 1
        return res

    def find(self, x):
        for i in range(4):
            a = x // self.m + self.dx[i]
            b = x % self.m + self.dy[i]
            if a < 0 or a >= self.n or b < 0 or b >= self.m or self.g[a][b]:
                continue

            j = a * self.m + b
            if not self.st[j]:
                self.st[j] = True
                if self.match[j] == -1 or self.find(self.match[j]):
                    self.match[j] = x
                    return True
        return False


if __name__ == "__main__":
    print(Solution().domino(2, 3, [[1, 0], [1, 1]]))  # 2
